File: skills/openclix-analytics/assets/ios/OpenClix/OpenClixAnalyticsEmitter.swift
import Foundation

public enum OpenClixSourceType: String {
    case app
    case system
}

public enum OpenClixAnalysisPeriod: String {
    case pre
    case post
}

public struct OpenClixAnalyticsEvent {
    public let name: String
    public let sourceType: OpenClixSourceType
    public let properties: [String: Any]

    public init(name: String, sourceType: OpenClixSourceType, properties: [String: Any] = [:]) {
        self.name = name
        self.sourceType = sourceType
        self.properties = properties
    }
}

public final class OpenClixAnalyticsEmitter {
    public typealias Sink = (_ eventName: String, _ properties: [String: Any]) -> Void
    public typealias NameTransform = (_ canonicalName: String) -> String

    private let platform: String
    private let analysisPeriod: OpenClixAnalysisPeriod
    private let campaignActive: Bool
    private let sink: Sink
    private let nameTransform: NameTransform?

    public init(
        platform: String,
        analysisPeriod: OpenClixAnalysisPeriod,
        campaignActive: Bool,
        sink: @escaping Sink,
        nameTransform: NameTransform? = nil
    ) {
        self.platform = platform
        self.analysisPeriod = analysisPeriod
        self.campaignActive = campaignActive
        self.sink = sink
        self.nameTransform = nameTransform
    }

    public func emit(_ event: OpenClixAnalyticsEvent) {
        var merged = event.properties
        merged["openclix_source"] = "openclix"
        merged["openclix_event_name"] = event.name
        merged["openclix_source_type"] = event.sourceType.rawValue
        merged["openclix_platform"] = platform
        merged["openclix_campaign_id"] = stringOrNil(event.properties["campaign_id"])
        merged["openclix_queued_message_id"] = stringOrNil(event.properties["queued_message_id"])
        merged["openclix_channel_type"] = stringOrNil(event.properties["channel_type"])
        merged["openclix_analysis_period"] = analysisPeriod.rawValue
        merged["openclix_campaign_active"] = campaignActive ? "true" : "false"

        let outboundName = nameTransform?(event.name) ?? event.name
        sink(outboundName, merged)
    }

    private func stringOrNil(_ value: Any?) -> String? {
        return value as? String
    }
}

public func normalizeFirebaseEventName(_ name: String) -> String {
    let lowered = name.lowercased()
    let normalized = lowered.replacingOccurrences(
        of: "[^a-z0-9_]",
        with: "_",
        options: .regularExpression
    )
    let prefixed = normalized.range(of: "^[a-z]", options: .regularExpression) == nil
        ? "oc_\(normalized)"
        : normalized
    return String(prefixed.prefix(40))
}

File: skills/openclix-analytics/assets/react-native/src/openclix/analytics/OpenClixAnalyticsEmitter.ts
export type OpenClixSourceType = 'app' | 'system';
export type OpenClixAnalysisPeriod = 'pre' | 'post';

export interface OpenClixAnalyticsEvent {
  name: string;
  sourceType: OpenClixSourceType;
  properties?: Record<string, unknown>;
}

export interface OpenClixAnalyticsEmitterConfig {
  platform: 'expo' | 'react-native';
  analysisPeriod: OpenClixAnalysisPeriod;
  campaignActive: boolean;
  sink: (eventName: string, properties: Record<string, unknown>) => Promise<void> | void;
  eventNameTransform?: (canonicalName: string) => string;
}

export class OpenClixAnalyticsEmitter {
  constructor(private readonly config: OpenClixAnalyticsEmitterConfig) {}

  async emit(event: OpenClixAnalyticsEvent): Promise<void> {
    const merged: Record<string, unknown> = {
      ...(event.properties ?? {}),
      openclix_source: 'openclix',
      openclix_event_name: event.name,
      openclix_source_type: event.sourceType,
      openclix_platform: this.config.platform,
      openclix_campaign_id: this.getString(event.properties, 'campaign_id'),
      openclix_queued_message_id: this.getString(event.properties, 'queued_message_id'),
      openclix_channel_type: this.getString(event.properties, 'channel_type'),
      openclix_analysis_period: this.config.analysisPeriod,
      openclix_campaign_active: this.config.campaignActive ? 'true' : 'false',
    };

    const outboundName = this.config.eventNameTransform
      ? this.config.eventNameTransform(event.name)
      : event.name;

    await this.config.sink(outboundName, merged);
  }

  private getString(source: Record<string, unknown> | undefined, key: string): string | null {
    const value = source?.[key];
    return typeof value === 'string' ? value : null;
  }
}

export function normalizeFirebaseEventName(name: string): string {
  let normalized = name.toLowerCase().replace(/[^a-z0-9_]/g, '_');
  if (!/^[a-z]/.test(normalized)) normalized = `oc_${normalized}`;
  return normalized.slice(0, 40);
}
